Here we describe the script/data type we used for each main figure. 

Please noted that we process our Micro-C, Hi-C, Capture Micro-C data based on 4DN Hi-C data processing pipeline (https://data.4dnucleome.org/resources/data-analysis/hi_c-processing-pipeline) to get the all the necessary files ```.hic```, ```.bam```, ```.cool```, ```.mcool``` required for different downstream softwares. If you are insterested to use any script here, pelase make sure you have the correct format of input files.

We processed the publically available ChIP-seq data achieved from GEO and ENCODE to recognize different regulatory elements (Promoter, Enhancer, Insulator, NDR, Repressed Regions etc.) in C42B. For ChIP-seq data, we processed them based on the pipeline from ENCODE. Please check https://www.encodeproject.org/chip-seq/histone/
For NOMe-seq, we got the C42B NOMe-seq data from previous study ([Rhie et al., 2018](https://pubmed.ncbi.nlm.nih.gov/29429977Bis/)). For NOMe-seq data processing, we used BSMAP ([Xi and Li 2009](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-10-232)) to align ```.fastq``` file to bisulfile-converted genome and Bis-SNP ([Liu et al. 2012](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-7-r61)) to identify meethylation status of CpG and GpC sites from BAM file.
